lib/improver/wind_direction.py

```diff
@@ -0,0 +1,237 @@
+# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+"""Module containing wind direction averaging plugins."""
+
+import iris
+import numpy as np
+from improver.utilities.cube_manipulation import enforce_float32_precision
+
+
+class WindDirection(object):
+    """Plugin to calculate average wind direction from ensemble members.
+
+    This holds the function to calculate the average wind direction
+    from ensemble members using a complex numbers approach.
+
+
+    """
+
+    def __init__(self):
+        """
+        Initialise class.
+        """
+        pass
+
+    def __repr__(self):
+        """Represent the configured plugin instance as a string."""
+        desc = ('<WindDirection>')
+        return desc
+
+    @staticmethod
+    def deg_to_complex(angle_deg, r=1):
+        """Takes an array or float containing an angle in degrees
+           and returns a complex equivalent.
+           The r value can be used to weigh values - but it is set
+           to 1 for now. """
+
+        # Convert from degrees to radians.
+        angle_rad = np.deg2rad(angle_deg)
+
+        # Derive real and imaginary components.
+        a = r*np.cos(angle_rad)
+        b = r*np.sin(angle_rad)
+
+        # Combine components into a complex number and return.
+        return a + 1j*b
+
+    @staticmethod
+    def complex_to_deg(complex_in):
+        """Take complex number and find angle in degrees from 360.
+           the "np.angle" function returns negative numbers when the input
+           is greater than 180. Therefore additional processing is needed
+           to ensure that the angle is between 0-359."""
+
+        angle = np.angle(complex_in, deg=True)
+
+        # If the input is an array then need to use native np.functions
+        # for speed rather than applying vector functions to the array.
+        if type(angle) is np.ndarray:
+
+            # If angle negative value - add 360 degrees.
+            angle = np.where(angle < 0, angle+360, angle)
+
+            # If angle == 360 - set to zero degrees.
+            # Due to floating point - need to round value before using
+            # equal operator.
+            round_angle = np.around(angle, 2)
+            angle = np.where(round_angle == 360, 0.0, angle)
+        else:
+            if angle < 0:
+                angle = angle + 360.0
+            if round(angle, 2) == 360.00:
+                angle = 0.0
+        return angle
+
+    @staticmethod
+    def find_r_values(complex_in):
```

Comment:
Do you need this function? At the moment these are set to 1, and in the future you'll have an array of r values that you can access from whatever way you calculate them from? What do you think?

Reply:
We do need this function!
So when we calculate the average of some complex values - the average is actually the midpoint between the two points in complex co-ordinates.
So if we have a wind direction at 0 and 180 degrees - the mid point is in the middle of the circle and therefore r=0.
So when we have r=0, we know that theres something wrong and then we can use the function "wind_dir_decider" to sort it out.